alphagov/support
# Add functionality to set Zendesk custom fields and form template

## Context

The current Content Change Request form also doesn't make the most of Zendesk. Making better use of custom fields would allow content design teams to improve internal reporting (e.g. improvement work tickets vs factual inaccuracy tickets).
The form template and custom fields were [created in the UI](https://support.zendesk.com/hc/en-us/articles/4408838961562).

[Trello card](https://trello.com/c/9j3DslWM/3176-apply-zendesk-mapping-to-the-content-requests-form-5)

## What

- Adds ability to set custom fields and form template in Zendesk
- Sets the following custom fields for Content change request tickets:
    - [CR] Reason for the request (required)
    - [CR] Subject Area(required)
    - [CR] Deadline
    - [CR] Do not publish before
    - [CR] Deadline (time)
    - [CR] Do not publish before (time)
- Sets 'Content Request' form template for Content change request tickets

| Before | After |
| --- | --- |
| <img width="581" alt="Screenshot 2023-08-24 at 10 23 52" src="https://github.com/alphagov/support/assets/38078064/53e9bde1-4922-47a1-a4e3-d378ff5dec19"> | <img width="607" alt="Screenshot 2023-08-23 at 15 16 11" src="https://github.com/alphagov/support/assets/38078064/480f2472-9d47-41c4-9f23-c127c6b04416"> |

Ticket examples:

- [Ticket without time constraint](https://govuk.zendesk.com/agent/tickets/5460403)
- [Ticket with time constraint](https://govuk.zendesk.com/agent/tickets/5460411)

## Review notes
- Best reviewed commit by commit (see message for more details)
- Since the custom fields information needs to be hardcoded, I decided to use a YAML file to store it. Loading ruby classes from a YAML file introduces a level of abstraction  but has the following benefits:
    - allows us to store all information about the Zendesk Custom fields in a YAML file (the app doesn't seem to have access to the custom fields end point via Zendesk API)
    - specific Ticket classes don't need to know about field type
    - CustomField class is not responsible for validating/formatting values of different types
    - makes it easy to extend should we want to populate other type of fields such as multi-select.

## How to test it
- deploy the branch to integration
- visit https://support.integration.publishing.service.gov.uk/content_change_request/new
- try different options
- find submitted ticket in [Zendesk](https://govuk.zendesk.com/agent/search/1) 
- try other forms

---
⚠️ This repo is Continuously Deployed: make sure you [follow the guidance](https://docs.publishing.service.gov.uk/manual/development-pipeline.html#merge-your-own-pull-request) ⚠️

Follow [these steps](https://guides.rubyonrails.org/upgrading_ruby_on_rails.html) if you are doing a Rails upgrade.
